### PR TITLE
Use iteration instead of recursion for `connect`

### PR DESCRIPTION
**Pull request by jyn514**
Look at this beautiful waste of time!

Closes https://github.com/algesten/ureq/issues/290.

### CODE REVIEW

File: src/response.rs
Comment (jsha) [inline src/response.rs:56]: This comment needs updating. In particular it should document which way the vector goes (first request to last request, or vice versa).

File: src/error.rs
Comment (jsha) [inline src/error.rs:106]: If you remove the `Hist` iterator, you'll want to change this `.last()` to `.first()`.

**Review by jsha — CHANGES_REQUESTED** [review]
Thanks for this! I like the general concept. A couple of quick thoughts:

The main thing we want is for the latest Response in a chain at any given point to have a history of the redirects that led to it. Right now, each response gets a cloned Vec. It might be sufficient for each Response to take the history Vec from the previous response and add one to it. That said, in the process of typing that out I realized it would make the relationship between one Request and the next a little more complicated, since we'd have to mutate the previous Request. So let's not do that.

We still have the `Hist` iterator type, but I don't think we need that anymore. That was just a hack to do iteration over a chain of Arcs. Now that we have a Vec, we can use Vec's iterator. We don't even strictly speaking need an iterator - but since Vec::iter() returns an ExactSizeIterator, using `.last()` is a perfectly efficient way to get the last item.

I'm not totally sold on the Retry concept, because it combines two somewhat unrelated purposes - retrying due to broken connection, and following a redirect. What do you think of leaving "retry due to broken connection" as a recursive operation (since it should never go more than one deep for any given Unit), and treating redirects as a more traditional loop?

For instance, we could have a `connect_inner` that returns `Result<Response, Transport>`, and `connect` would have a loop that calls `connect_inner`, and if the `Response` is a 3xx, it performs the rI don't totally like edirect logic and repeats the loop.

File: src/response.rs
Comment (jyn514) [reply src/response.rs:56]: `hist.iter()` returns a reverse iterator. I agree this is confusing, I'll change it to return a normal iterator and use `first()` here instead.

**Review by jyn514 — COMMENTED** [review]
> Right now, each response gets a cloned Vec.

The vec is not cloned. This is the reason I moved `previous` out of do_from_request, so I would be able to access it directly even in case of an IO error.

> What do you think of leaving "retry due to broken connection" as a recursive operation (since it should never go more than one deep for any given Unit)

Why will it never go more than one deep? I don't see any state tracking that.

Assuming it never goes more than one deep, I agree separating the two makes sense.

> For instance, we could have a connect_inner that returns Result<Response, Transport>, and connect would have a loop that calls connect_inner, and if the Response is a 3xx, it performs the redirect logic and repeats the loop.

As in, move the 300 handling out of `connect_inner`? Yeah, that sounds a lot simpler.

File: src/response.rs
Comment (jsha) [inline src/response.rs]: > > Right now, each response gets a cloned Vec.

> The vec is not cloned. This is the reason I moved previous out of do_from_request, so I would be able to access it directly even in case of an IO error.

Isn't this (line 463) cloning the vec?

Also, I noticed this is still taking an `Arc<Response>`, but I think it can just be a `&Response` now, right?

**Review by jsha — COMMENTED** [review]
> Why will it never go more than one deep? I don't see any state tracking that.

[RFC 7230 says this](https://tools.ietf.org/html/rfc7230#section-6.3.1):

> A client SHOULD NOT automatically retry a failed automatic retry.

This is implemented by the `use_pooled` parameter to `connect`. When we first arrive, it's true. If we do an automatic retry, it's false on the second call. We also gate both retry paths on `is_recycled`, which can only be true if we asked for a pooled connection and got one. This is definitely a bit convoluted and spread over more locations than it should be. We should clarify the code.

File: src/response.rs
Comment (jyn514) [reply src/response.rs]: This is cloning the vec, but it's for tests only, because it takes a `Vec`. The main logic in `connect_inner` doesn't do any clones.

> Also, I noticed this is still taking an Arc<Response>, but I think it can just be a &Response now, right?

I think so, yes. I'll try changing that.

File: src/unit.rs
Comment (jsha) [inline src/unit.rs:168]: This is now state entirely internal to the `connect` fn, so it can be moved to a local variable.

**Review by jsha — CHANGES_REQUESTED** [review]
I'm finding the ownership flow of `previous` back and forth between the two functions to be fairly confusing. Is it possible for ownership of `previous` to remain entirely within the outer `connect` fn?

**Comment by jyn514** [comment]
> Is it possible for ownership of previous to remain entirely within the outer connect fn?

Yes, I think so - it just needs to be set before returning from `connect()`, it doesn't matter very much what `resp.history` is before then.

File: src/unit.rs
Comment (jyn514) [reply src/unit.rs:168]: It's not entirely internal state - `send_prelude` needs to know if this is a redirect or not. But it can do that just as easily with a slice instead.

Comment (jyn514) [reply src/unit.rs:168]: Oh wait, I just realized you meant this should only ever be called with `vec![]`. Yes, it can be removed from `connect`, just not `connect_inner`.

**Comment by jyn514** [comment]
I'm pretty happy with this :) let me know when/if I should squash the commits.

**Review by jsha — APPROVED** [review]
Yeah, this is looking really nice! I think this is ready to go. No need to squash the commits. When I go to merge, I'll use GitHub's "squash and merge" function.

I'll come back and give it a second look with fresh eyes in the morning before hitting the button.

**Comment by algesten** [comment]
Wow. This is quite a PR. Thanks! I'll go through it later today too.

File: src/unit.rs
Comment (jsha) [inline src/unit.rs:169]: Nit: I'd like a type annotation on this. Otherwise you have to read pretty far to figure out what's stored in this vec.

File: src/unit.rs
Comment (jsha) [inline src/unit.rs:175]: Nit: I'd like to reorder the cases in this loop to favor early returns and reduce indentation. Specifically, something like:

```rust
if !(300..399).contains(&resp.status()) || unit.agent.config.redirects == 0 {
  return Ok(resp);
}
... otherwise, handle a redirect ...
```

This also allows getting rid of the `continue`, which IMO makes this feel more like a natural loop, where falling through to the bottom means the loop continues by default.

File: src/unit.rs
Comment (jsha) [inline src/unit.rs:173]: I think this is slightly early to push the unit's URL onto `previous`. Doesn't this mean a resp with no redirects contains itself in its history? We should make sure we have tests that catch this, if it's the case (I might also be misunderstanding here).

**Review by jsha — CHANGES_REQUESTED** [review]

File: src/unit.rs
Comment (jyn514) [reply src/unit.rs:169]: I'm confused why you say that. `history.push(unit.url.to_string())` is only a couple lines below: https://github.com/algesten/ureq/blob/de669f65eca0d3d5a19b6b750449760e8a50f5ae/src/unit.rs#L172

File: src/unit.rs
Comment (jyn514) [reply src/unit.rs]: Done, although I used `break` instead of return to avoid repeating the `Ok` a lot.

I also noticed that this returns `Ok(resp)` when `redirects == 0`, but `Err(TooManyRedirects)` for any other value of redirects. Is that intentional?

File: src/unit.rs
Comment (jyn514) [reply src/unit.rs]: Ooh, excellent catch. Yes, this should only push the history if we're sure we're going to redirect (near the bottom, after calculating `new_method`.

That brings up another issue: I don't think the history tests actually test anything. Because they add items to the history manually, it doesn't matter what `connect` does, they'll pass no matter what.

**Review by jsha — APPROVED** [review]
Looks good to me! @algesten, since you expressed interest in reviewing, I'll wait on your approval.

File: src/unit.rs
Comment (jsha) [reply src/unit.rs]: Yep, that's intentional. `0` is a special value that means "turn off redirect handling."

File: src/unit.rs
Comment (jsha) [reply src/unit.rs]: Heh, true. I got hung up on the line before that:

https://github.com/algesten/ureq/blob/de669f65eca0d3d5a19b6b750449760e8a50f5ae/src/unit.rs#L171

And thought "I wonder what is being passed to connect_inner? I guess I'd have to go look at the definition for that function.

In general I've been developing a preference for explicit type annotations in a lot of cases. But I won't be picky about it in this specific spot.

**Review by algesten — APPROVED** [review]
Yep. This all looks good to me. Merge away!

**Pull request merged**
